File: companies/Cashfree/229.majority-element-ii.go
package leetcode

// MajorityElement229 is Boyer-Moore generalized for elements appearing more than
// n/3 times. There can be at most two candidates, so a second pass verifies the
// actual counts.
//
// Time: O(n). Space: O(1).
func MajorityElement229(nums []int) []int {
	cand1, cand2, count1, count2 := 0, 1, 0, 0
	for _, x := range nums {
		if count1 > 0 && x == cand1 {
			count1++
		} else if count2 > 0 && x == cand2 {
			count2++
		} else if count1 == 0 {
			cand1, count1 = x, 1
		} else if count2 == 0 {
			cand2, count2 = x, 1
		} else {
			count1--
			count2--
		}
	}
	c1, c2 := 0, 0
	for _, x := range nums {
		if count1 > 0 && x == cand1 {
			c1++
		}
		if count2 > 0 && x == cand2 {
			c2++
		}
	}
	ans := []int{}
	if count1 > 0 && c1 > len(nums)/3 {
		ans = append(ans, cand1)
	}
	if count2 > 0 && cand2 != cand1 && c2 > len(nums)/3 {
		ans = append(ans, cand2)
	}
	return ans
}

File: companies/Cashfree/229.majority-element-ii.java
import java.util.*;

/**
 * Algorithm:
 * Boyer-Moore voting generalized for elements appearing more than n/3 times.
 * At most two such elements can exist, so keep two candidates and verify them
 * with a second pass.
 *
 * Complexity:
 * Time O(n), space O(1) excluding output.
 */
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        Integer cand1 = null;
        Integer cand2 = null;
        int count1 = 0;
        int count2 = 0;
        for (int num : nums) {
            if (cand1 != null && num == cand1) {
                count1++;
            } else if (cand2 != null && num == cand2) {
                count2++;
            } else if (count1 == 0) {
                cand1 = num;
                count1 = 1;
            } else if (count2 == 0) {
                cand2 = num;
                count2 = 1;
            } else {
                count1--;
                count2--;
            }
        }
        int actual1 = 0;
        int actual2 = 0;
        for (int num : nums) {
            if (cand1 != null && num == cand1) {
                actual1++;
            } else if (cand2 != null && num == cand2) {
                actual2++;
            }
        }
        List<Integer> ans = new ArrayList<>();
        if (cand1 != null && actual1 > nums.length / 3) {
            ans.add(cand1);
        }
        if (cand2 != null && actual2 > nums.length / 3) {
            ans.add(cand2);
        }
        return ans;
    }
}

